src/lib/bootloader/stage1.rb

```diff
@@ -196,203 +218,249 @@ def available_partitions(res)
-        self.generic_mbr = false
-      # handle diskless setup, in such case do not write boot code anywhere (bnc#874466)
-      # we need to detect what is mount on /boot and if it is nfs, then just
-      # skip this proposal. In other case if it is not nfs, then it is error and raise exception
-      # powernv do not have prep partition, so we do not have any partition to activate (bnc#970582)
-      elsif Yast::BootStorage.disk_with_boot_partition == "/dev/nfs" || Yast::Arch.board_powernv
-        self.activate = false
-        self.generic_mbr = false
-        return
-      else
-        raise "there is no prep partition"
-      end
+    def initialize(stage1)
+      @stage1 = stage1
     end
 
-    def proposed_prep_partition
-      partitions = Yast::BootStorage.prep_partitions
+    # Set "boot_*" flags in the globals map according to the boot device selected
+    # with parameter selected_location. Only a single boot device can be selected
+    # with this function. The function cannot be used to set a custom boot device.
+    # It will always be deleted.
+    #
+    # FIXME: `mbr_md is probably unneeded; AFA we can see, this decision is
+    # automatic anyway and perl-Bootloader should be able to make it without help
+    # from the user or the proposal.
+    #
+    # @param [Symbol, Array] selected_location symbol one of :boot, :root, :mbr,
+    #   :extended, `none or Array with first value :custom and second device for
+    #   custom devices
+    def assign_bootloader_device(selected_location)
+      log.info "assign bootloader device '#{selected_location.inspect}'"
+      # first, default to all off:
+      stage1.clear_devices
 
-      created = partitions.find do |part|
-        part_map = Yast::Storage.GetPartition(Yast::Storage.GetTargetMap, part)
-        part_map["create"] == true
-      end
+      case selected_location
+      when :root then stage1.add_udev_device(Yast::BootStorage.RootPartitionDevice)
+      when :boot then stage1.add_udev_device(Yast::BootStorage.BootPartitionDevice)
+      when :extended then stage1.add_udev_device(extended_partition)
+      when :mbr then stage1.add_udev_device(Yast::BootStorage.mbr_disk)
+      when :none then log.info "Resetting bootloader device"
+      when Array
+        if selected_location.first != :custom
+          raise "Unknown value to select bootloader device #{selected_location.inspect}"
+        end
 
-      if created
-        log.info "using freshly created prep partition #{created}"
-        return created
+        stage1.model.add_device(selected_location[1]) # add directly proposed value without changes
+      else
+        raise "Unknown value to select bootloader device #{selected_location.inspect}"
       end
+    end
 
-      same_disk_part = partitions.find do |part|
-        disk = Yast::Storage.GetDiskPartition(part)["disk"]
-        Yast::BootStorage.disk_with_boot_partition == disk
+    # x86_64 specific stage1 proposal
+    class X64 < Stage1Proposal
```

Comment:
so you want nested class in own file? so structure like lib/bootloader/stage1_proposal/x64.rb?